psi4/9b60184.nix

```diff
@@ -0,0 +1,244 @@
+
+    enableParallelBuilding = true;
+
+    configurePhase = ''
+      cmake -Bbuild ${toString specialInstallCmakeFlags}
+      cd build
+    '';
+  });
+
+
+in
+  stdenv.mkDerivation rec {
+    pname = "psi4";
+    version = "1.3.2";
+
+    nativeBuildInputs = [
+      cmake
+      perl
+      gfortran
+      makeWrapper
+      pkg-config
+    ];
+
+    buildInputs = [
+      gau2grid-2_0_4
+      libint1_
+      libxc
+      blas_
+      lapack_
+      dkh_
+      pcmsolver_
+      libefp
+      chemps2_
+      hdf5
+      hdf5-cpp
+    ];
+
+    propagatedBuildInputs = [
+      pybind11
+      qcelemental
+      qcengine
+      numpy
+      pylibefp
+      deepdiff
+      dftd3
+      chemps2_
+    ] ++ qcelemental.passthru.requiredPythonModules
+      ++ qcengine.passthru.requiredPythonModules
+    ;
+
+    src = fetchFromGitHub  {
+      owner = "psi4";
+      repo = pname;
+      rev = "9b60184c5d161e4871c91ce29a44e3ac2c2a438e";
+      sha256= "1vh8dp3nw4fk1mnfv0w8ici1lzxyfn5han7hipqzsfxl75w76r18";
+    };
+
+    cmakeFlags = [
+      "-DDCMAKE_FIND_USE_SYSTEM_PACKAGE_REGISTRY=OFF"
+      "-DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF"
+      "-DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON"
+      "-DCMAKE_INSTALL_PREFIX=$out"
+      "-DBUILD_SHARED_LIBS=ON"
+      "-DENABLE_XHOST=OFF"
+      "-DENABLE_OPENMP=ON"
+      # gau2grid
+      "-DCMAKE_INSIST_FIND_PACKAGE_gau2grid=ON"
+      "-Dgau2grid_DIR=${gau2grid-2_0_4}/share/cmake/gau2grid"
+      # libint
+      "-DCMAKE_INSIST_FIND_PACKAGE_Libint=ON"
+      "-DLibint_DIR=${libint1_}/share/cmake/Libint"
+      # libxc
+      "-DCMAKE_INSIST_FIND_PACKAGE_Libxc=ON"
+      "-DLibxc_DIR=${libxc_}/share/cmake/Libxc"
+      # pcmsolver
+      "-DCMAKE_INSIST_FIND_PACKAGE_PCMSolver=ON"
+      "-DENABLE_PCMSolver=ON"
+      "-DPCMSolver_DIR=${pcmsolver_}/share/cmake/PCMSolver"
+      # DKH
+      "-DENABLE_dkh=ON"
+      "-DCMAKE_INSIST_FIND_PACKAGE_dkh=ON"
+      "-Ddkh_DIR=${dkh_}/share/cmake/dkh"
+      # LibEFP
+      "-DENABLE_libefp=ON"
+      # CheMPS2
+      "-DENABLE_CheMPS2=ON"
+      # Prefix path for all external packages
+      "-DCMAKE_PREFIX_PATH=\"${gau2grid-2_0_4};${libint1_};${libxc_};${qcelemental};${pcmsolver_};${dkh_};${libefp};${chemps2_}\""
+    ];
+
+    enableParallelBuilding = true;
+
+    configurePhase = ''
+      cmake -Bbuild ${toString cmakeFlags} -DCMAKE_INSTALL_PREFIX=$out
+      cd build
+    '';
+
+    binSearchPath = with stdenv.lib; strings.makeSearchPath "bin" [
+      dftd3
+    ];
+
+    libSearchPath = with stdenv.lib; strings.makeSearchPath "lib/${python.executable}/site-packages" ([
+      pybind11
+      qcelemental
+      numpy
+      libefp
+      deepdiff
+    ] ++ qcelemental.passthru.requiredPythonModules
+    );
+
+    postFixup = ''
+      # Make libraries and external binaries available
+      wrapProgram $out/bin/psi4 \
+        --prefix PATH : ${binSearchPath} \
+        --prefix PYTHONPATH : ${libSearchPath} \
```

Comment:
This could be done shorter by using `python.withPackages`. Here's the basic idea:
```
 buildInputs = [ python' ];
 python' = python.withPackages(ps: with ps; [ numpy  pybind11 qcelemental libefp deepdiff])
```
That way you get a wrapped python interpreter that has the packages already in its `PYTHONPATH`.

Reply:
I would prefer not to do this :slightly_smiling_face:. It actually caused some problems in our applications. Psi4 does not simply require python modules at runtime, but is actually a python module, that can be imported with `import psi4` in python. A colleague writes a python program - [pysisyphus](https://github.com/eljost/pysisyphus) - implementing advanced optimisers. It wraps several QC packages and for Psi4 it uses the python library. Now when we build the derivation of pysisyphus with nix and Psi4 is one of the `propagatedBuildInputs` to pysis and one uses the `python.withPackages` stuff, the derivations fail to build, as `python.withPackages` seems to create an extra derivation with symlinkJoin or something? In the end there are multiple versions of shared dependencies like numpy and python gets confused.

Long story short: i think the `python.withPackages` version causes some problems, when using Psi4 as a python module, rather than as an executable.

Reply:
Same as above:
Oh I did not realise that is defined as python package. In that case: did you ever try to build it with buildPythonPackage instead of mkDerivation? That should take of most the python environment issues.